boot/flags.go

```diff
@@ -0,0 +1,282 @@
+					disallowedFlags = append(disallowedFlags, flag)
+				}
+			}
+		}
+	}
+	if len(allowedFlags) != len(flags) {
+		return allowedFlags, unknownFlagErr(fmt.Sprintf("unknown boot flags %v not allowed", disallowedFlags))
+	}
+	return flags, nil
+}
+
+func serializeBootFlags(flags []string) string {
+	// drop empty strings before serializing
+	nonEmptyFlags := make([]string, 0, len(flags))
+	for _, flag := range flags {
+		if strings.TrimSpace(flag) != "" {
+			nonEmptyFlags = append(nonEmptyFlags, flag)
+		}
+	}
+
+	return strings.Join(nonEmptyFlags, ",")
+}
+
+// setImageBootFlags writes the provided flags to the bootenv of the recovery
+// bootloader in the specified system rootDir. It is only meant to be called at
+// prepare-image customization time by ubuntu-image/prepare-image.
+func setImageBootFlags(flags []string, rootDir string) error {
+	// check that the flagList is supported
+	if _, err := checkBootFlagList(flags, understoodBootFlags); err != nil {
+		return err
+	}
+
+	// also ensure that the serialized value of the boot flags fits inside the
+	// bootenv value, on lk systems the max size of a bootenv value is 255 chars
+	s := serializeBootFlags(flags)
+	if len(s) > 254 {
+		return fmt.Errorf("internal error: boot flags too large to fit inside bootenv value")
+	}
+
+	// now find the recovery bootloader in the system dir and set the value on
+	// it
+	opts := &bootloader.Options{
+		Role: bootloader.RoleRecovery,
+	}
+	bl, err := bootloader.Find(rootDir, opts)
+	if err != nil {
+		return err
+	}
+
+	return bl.SetBootVars(map[string]string{
+		"snapd_boot_flags": s,
+	})
+}
+
+// InitramfsActiveBootFlags returns the set of boot flags that are currently set
+// for the current boot, by querying them directly from the source. This method
+// is only meant to be used from the initramfs, since it may query the bootenv
+// or query the modeenv depending on the current mode of the system.
+// For detecting the current set of boot flags outside of the initramfs, use
+// BootFlags(), which will query for the runtime version of the flags in /run
+// that the initramfs will have setup for userspace.
+// Note that no filtering is done on the flags in order to allow new flags to be
+// used by a userspace that is newer than the initramfs, but empty flags will be
+// dropped automatically.
+// Only to be used on UC20+ systems with recovery systems.
+func InitramfsActiveBootFlags(mode string) ([]string, error) {
+	switch mode {
+	case ModeRecover:
+		// no boot flags are consumed / used on recover mode, so return nothing
+		return nil, nil
+
+	case ModeRun:
+		// boot flags come from the modeenv
+		modeenv, err := ReadModeenv(InitramfsWritableDir)
+		if err != nil {
+			return nil, err
+		}
+
+		// TODO: consider passing in the modeenv or returning the modeenv here
+		// to reduce the number of times we read the modeenv ?
+		return modeenv.BootFlags, nil
+
+	case ModeInstall:
+		// boot flags always come from the bootenv for the recover bootloader
```

Comment:
```suggestion
		// boot flags always come from the bootenv of the recovery bootloader
```

Reply:
I did this here since I needed a master merge to unbreak arch